core/common/src/main/java/alluxio/grpc/GrpcConnectionPool.java

```diff
@@ -0,0 +1,378 @@
+ * This class is used internally by {@link GrpcChannelBuilder} and {@link GrpcChannel}.
+ */
+@ThreadSafe
+public class GrpcConnectionPool {
+  private static final Logger LOG = LoggerFactory.getLogger(GrpcConnectionPool.class);
+
+  // Singleton instance.
+  private static GrpcConnectionPool sInstance;
+
+  static {
+    sInstance = new GrpcConnectionPool();
+  }
+
+  /** gRPC Managed channels/connections. */
+  private ConcurrentMap<GrpcConnectionKey, CountingReference<ManagedChannel>> mChannels;
+
+  /** Event loops. */
+  private ConcurrentMap<GrpcChannelKey, CountingReference<EventLoopGroup>> mEventLoops;
+
+  /** Used to assign order within a network group. */
+  private ConcurrentMap<GrpcChannelKey.NetworkGroup, AtomicLong> mNetworkGroupCounters;
+
+  /**
+   * Creates a new {@link GrpcConnectionPool}.
+   */
+  public GrpcConnectionPool() {
+    mChannels = new ConcurrentHashMap<>();
+    mEventLoops = new ConcurrentHashMap<>();
+    // Initialize counters for known network-groups.
+    mNetworkGroupCounters = new ConcurrentHashMap<>();
+    for (GrpcChannelKey.NetworkGroup group : GrpcChannelKey.NetworkGroup.values()) {
+      mNetworkGroupCounters.put(group, new AtomicLong());
+    }
+  }
+
+  /**
+   * @return the singleton pool instance
+   */
+  public static GrpcConnectionPool INSTANCE() {
+    return sInstance;
+  }
+
+  /**
+   * Acquires and increases the ref-count for the {@link ManagedChannel}.
+   *
+   * @param channelKey the channel key
+   * @param conf the Alluxio configuration
+   * @return a {@link GrpcConnection}
+   */
+  public GrpcConnection acquireConnection(GrpcChannelKey channelKey, AlluxioConfiguration conf) {
+    // Get a connection key.
+    GrpcConnectionKey connectionKey = getConnectionKey(channelKey, conf);
+    // Acquire connection.
+    CountingReference<ManagedChannel> connectionRef =
+        mChannels.compute(connectionKey, (key, ref) -> {
+          boolean shutdownExistingConnection = false;
+          int existingRefCount = 0;
+          if (ref != null) {
+            // Connection exists, wait for health check.
+            if (waitForConnectionReady(ref.get(), conf)) {
+              LOG.debug("Acquiring an existing connection. ConnectionKey: {}. Ref-count: {}", key,
+                  ref.getRefCount());
+
+              return ref.reference();
+            } else {
+              // Health check failed.
+              shutdownExistingConnection = true;
+            }
+          }
+          // Existing connection should shut-down.
+          if (shutdownExistingConnection) {
+            // TODO(ggezer): Implement GrpcConnectionListener for receiving notification.
+            existingRefCount = ref.getRefCount();
+            LOG.debug("Shutting down an existing unhealthy connection. "
+                + "ConnectionKey: {}. Ref-count: {}", key, existingRefCount);
+            // Shutdown the channel forcefully as it's already unhealthy.
+            shutdownManagedChannel(ref.get(), conf);
```

Comment:
Instead of shutting down here, should we let other users of the channel eventually fail and only shutdown after the refcount is zero?

Reply:
Though what would happen to the one that figured out the health issue?
- Assign a new group id and retry?
   - What if max connection is 1, so we keep retrying on the same channel?
- Block this one until other clients are failed released the connection?
  - This could never happen as they will be cached in pools and may not be used for a long time. 
    So this block could take at least as long as pool GC interval. (We can implement GrpcConnectionLister interface that is used to synchronously let them owners know)

Reply:
Oh I see, we cannot replace the value in the map because the other users don't have a direct reference to the channel.